File: android/app/src/main/kotlin/com/swastricare/health/data/repository/FamilyAlertPreferencesRepository.kt
package com.swastricare.health.data.repository

import android.util.Log
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for the `family_alert_preferences` table.
 *
 * Each row stores a caregiver's per-target alert preferences (which alert types are enabled,
 * quiet hours, missed-medication grace period). Primary key is the composite
 * (caregiver_user_id, target_health_profile_id).
 *
 * RLS restricts all operations to rows where `caregiver_user_id = auth.uid()`, so calls
 * require an authenticated session and the caller can only manage their own preferences.
 */
@Serializable
data class FamilyAlertPreferences(
    @SerialName("caregiver_user_id") val caregiverUserId: String,
    @SerialName("target_health_profile_id") val targetHealthProfileId: String,
    @SerialName("missed_medication_alerts") val missedMedicationAlerts: Boolean = true,
    @SerialName("low_hydration_alerts") val lowHydrationAlerts: Boolean = true,
    @SerialName("missed_vitals_alerts") val missedVitalsAlerts: Boolean = false,
    @SerialName("custom_nudge_alerts") val customNudgeAlerts: Boolean = true,
    /** "HH:mm:ss" 24-hour, nullable when quiet hours are disabled. */
    @SerialName("quiet_hours_start") val quietHoursStart: String? = null,
    @SerialName("quiet_hours_end") val quietHoursEnd: String? = null,
    /** DB CHECK constraint: 5..240. */
    @SerialName("missed_med_grace_minutes") val missedMedGraceMinutes: Int = 30,
)

@Singleton
class FamilyAlertPreferencesRepository @Inject constructor(
    private val supabaseClient: SupabaseClient,
) {
    companion object {
        private const val TAG = "FamilyAlertPrefsRepo"
        private const val TABLE = "family_alert_preferences"
        private const val ON_CONFLICT = "caregiver_user_id,target_health_profile_id"
    }

    /**
     * Fetch the caregiver's preferences for a specific target profile, or null if none saved yet.
     * Callers should treat null as "use defaults" — do not pre-create a row on first read.
     */
    suspend fun get(
        caregiverUserId: String,
        targetProfileId: String,
    ): Result<FamilyAlertPreferences?> = runCatching {
        supabaseClient.from(TABLE)
            .select {
                filter {
                    eq("caregiver_user_id", caregiverUserId)
                    eq("target_health_profile_id", targetProfileId)
                }
                limit(1)
            }
            .decodeSingleOrNull<FamilyAlertPreferences>()
    }.onFailure { e ->
        Log.e(TAG, "Failed to load alert preferences", e)
    }

    /**
     * Insert-or-update the preferences row. Safe to call on every save — `updated_at` is
     * bumped by a DB trigger.
     */
    suspend fun upsert(prefs: FamilyAlertPreferences): Result<Unit> = runCatching {
        supabaseClient.from(TABLE).upsert(prefs, onConflict = ON_CONFLICT)
        Unit
    }.onFailure { e ->
        Log.e(TAG, "Failed to upsert alert preferences", e)
    }
}

File: swastricare-mobile-swift/Services/Family/FamilyAlertPreferencesRepository.swift
//
//  FamilyAlertPreferencesRepository.swift
//  swastricare-mobile-swift
//
//  Family Monitoring — per-(caregiver, target_health_profile_id) alert preferences.
//  Mirrors the Android FamilyAlertPreferencesRepository.
//

import Foundation
import Supabase

// MARK: - Model

struct FamilyAlertPreferences: Codable, Sendable, Equatable {
    let caregiverUserId: String
    let targetHealthProfileId: String
    var missedMedicationAlerts: Bool
    var lowHydrationAlerts: Bool
    var missedVitalsAlerts: Bool
    var customNudgeAlerts: Bool
    var quietHoursStart: String?       // "HH:mm:ss"
    var quietHoursEnd: String?         // "HH:mm:ss"
    var missedMedGraceMinutes: Int

    init(
        caregiverUserId: String,
        targetHealthProfileId: String,
        missedMedicationAlerts: Bool = true,
        lowHydrationAlerts: Bool = true,
        missedVitalsAlerts: Bool = false,
        customNudgeAlerts: Bool = true,
        quietHoursStart: String? = nil,
        quietHoursEnd: String? = nil,
        missedMedGraceMinutes: Int = 30
    ) {
        self.caregiverUserId = caregiverUserId
        self.targetHealthProfileId = targetHealthProfileId
        self.missedMedicationAlerts = missedMedicationAlerts
        self.lowHydrationAlerts = lowHydrationAlerts
        self.missedVitalsAlerts = missedVitalsAlerts
        self.customNudgeAlerts = customNudgeAlerts
        self.quietHoursStart = quietHoursStart
        self.quietHoursEnd = quietHoursEnd
        self.missedMedGraceMinutes = missedMedGraceMinutes
    }

    enum CodingKeys: String, CodingKey {
        case caregiverUserId = "caregiver_user_id"
        case targetHealthProfileId = "target_health_profile_id"
        case missedMedicationAlerts = "missed_medication_alerts"
        case lowHydrationAlerts = "low_hydration_alerts"
        case missedVitalsAlerts = "missed_vitals_alerts"
        case customNudgeAlerts = "custom_nudge_alerts"
        case quietHoursStart = "quiet_hours_start"
        case quietHoursEnd = "quiet_hours_end"
        case missedMedGraceMinutes = "missed_med_grace_minutes"
    }
}

// MARK: - Protocol

protocol FamilyAlertPreferencesRepositoryProtocol: Sendable {
    func get(caregiverUserId: String, targetProfileId: String) async throws -> FamilyAlertPreferences?
    func upsert(_ prefs: FamilyAlertPreferences) async throws
}

// MARK: - Implementation

final class FamilyAlertPreferencesRepository: FamilyAlertPreferencesRepositoryProtocol, @unchecked Sendable {
    static let shared = FamilyAlertPreferencesRepository()

    private let client: SupabaseClient

    private init() {
        self.client = SupabaseManager.shared.client
    }

    func get(caregiverUserId: String, targetProfileId: String) async throws -> FamilyAlertPreferences? {
        let rows: [FamilyAlertPreferences] = try await client
            .from("family_alert_preferences")
            .select()
            .eq("caregiver_user_id", value: caregiverUserId)
            .eq("target_health_profile_id", value: targetProfileId)
            .limit(1)
            .execute()
            .value
        return rows.first
    }

    func upsert(_ prefs: FamilyAlertPreferences) async throws {
        try await client
            .from("family_alert_preferences")
            .upsert(prefs, onConflict: "caregiver_user_id,target_health_profile_id")
            .execute()
    }
}
